lib/handler/status/durations.c

```diff
@@ -0,0 +1,231 @@
+            DURATION_FMT(body-time) ","
+            DURATION_FMT(request-total-time) ","
+            DURATION_FMT(process-time) ","
+            DURATION_FMT(response-time) ","
+            DURATION_FMT(duration),
+#undef DURATION_FMT
+#define DURATION_VALS(x) gkc_query(agg_stats->stats.x, 0), gkc_query(agg_stats->stats.x, 0.25), \
+                        gkc_query(agg_stats->stats.x, 0.5), gkc_query(agg_stats->stats.x, 0.75), \
+                        gkc_query(agg_stats->stats.x, 0.99)
+
+                        DURATION_VALS(connect_time), DURATION_VALS(header_time),
+                        DURATION_VALS(body_time), DURATION_VALS(request_total_time),
+                        DURATION_VALS(process_time), DURATION_VALS(response_time),
+                        DURATION_VALS(duration));
+#undef DURATION_VALS
+#undef BUFSIZE
+
+    duration_stats_free(&agg_stats->stats);
+    pthread_mutex_destroy(&agg_stats->mutex);
+
+    free(agg_stats);
+    return ret;
+}
+
+static int64_t diff_timeval(struct timeval *from, struct timeval *until)
+{
+    int32_t delta_sec = (int32_t)until->tv_sec - (int32_t)from->tv_sec;
+    int32_t delta_usec = (int32_t)until->tv_usec - (int32_t)from->tv_usec;
+    int64_t delta = (int64_t)(delta_sec * 1000*1000L) + delta_usec;
+
+    return delta;
+}
+
+static inline int timeval_is_null(struct timeval *tv)
+{
+        return tv->tv_sec == 0;
+}
+
+static void stat_access(h2o_logger_t *_self, h2o_req_t *req)
+{
+    struct st_duration_stats_t *ctx_stats = h2o_context_get_logger_context(req->conn->ctx, _self);
+#define ADD_OBSERVATION(x, from, until) \
+    do { \
+        /* skip zero values: this happens in h2 if a request is reset \
+         * before being sent */ \
+        if (!((from)->tv_sec + (from)->tv_usec)) { \
+            break; \
+        } \
+        if (!((until)->tv_sec + (until)->tv_usec)) { \
+            break; \
+        } \
+        int64_t dur = diff_timeval((from), (until)); \
+        gkc_insert_value(ctx_stats->x, dur); \
+    } while(0)
+
+    ADD_OBSERVATION(connect_time, &req->conn->connected_at, &req->timestamps.request_begin_at);
+    ADD_OBSERVATION(header_time, &req->timestamps.request_begin_at, timeval_is_null(&req->timestamps.request_body_begin_at)
+            ? &req->processed_at.at
+            : &req->timestamps.request_body_begin_at);
+    ADD_OBSERVATION(body_time, timeval_is_null(&req->timestamps.request_body_begin_at)
+            ? &req->processed_at.at
+            : &req->timestamps.request_body_begin_at, &req->processed_at.at);
+    ADD_OBSERVATION(request_total_time, &req->timestamps.request_begin_at, &req->processed_at.at);
+    ADD_OBSERVATION(process_time, &req->processed_at.at, &req->timestamps.response_start_at);
+    ADD_OBSERVATION(response_time, &req->timestamps.response_start_at, &req->timestamps.response_end_at);
+    ADD_OBSERVATION(duration, &req->timestamps.request_begin_at, &req->timestamps.response_end_at);
+#undef ADD_OBSERVATION
```

Comment:
Is my understanding correct that the logic of calculating these numbers is a duplication of that found in `lib/core/logconf.c`?

Assuming that it is correct, how about creating inline functions in `h2o.h` that returns various durations (or a _null_ in some form), and use them from both logconf.c and durations.c?